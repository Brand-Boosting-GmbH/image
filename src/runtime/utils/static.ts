const manifest = {}

export function useManifest (nuxtContext?) {
  if (process.client) {
    if (typeof window.$nuxt !== 'undefined') {
      const pageImages = (window.$nuxt as any)._pagePayload?.data?.[0]?._img || {}
      Object.assign(manifest, pageImages)
      return manifest
    }

    // merge new manifests on route change
    if (nuxtContext) {
      nuxtContext.app.router.afterEach(() => useManifest())
    }

    // make sure manifest is initialized
    (window as any).onNuxtReady(useManifest)
  }
  return manifest
}
