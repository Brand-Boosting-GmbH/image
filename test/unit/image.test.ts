/**
 * @jest-environment jsdom
 */

import type Vue from 'vue'
import { Wrapper } from '@vue/test-utils'

import { getSrc, mountWithImg } from './utils/mount'
import { mockObserver } from './utils/observer'
import { nextTick } from './utils/tick'

import NuxtImg from '~/runtime/components/nuxt-img.vue'

describe('Renders simple image', () => {
  let wrapper: Wrapper<Vue>
  const src = '/image.png'

  beforeEach(() => {
    wrapper = mountWithImg(NuxtImg, {
      width: 200,
      height: 200,
      sizes: '200,500:500,900:900',
      src
    })
  })

  test('Matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test.todo('alt attribute is generated')
  // () => {
  //   expect((wrapper.vm as any).generatedAlt).toEqual('image')
  //   const domAlt = wrapper.element.getAttribute('alt')
  //   expect(domAlt).toEqual('image')
  // }

  test('props.src is picked up by getImage()', () => {
    const domSrc = wrapper.element.getAttribute('src')
    expect(domSrc).toEqual(getSrc(src))
  })

  test('props.src is reactive', (done) => {
    const newSource = '/image.jpeg'
    wrapper.setProps({ src: newSource })
    process.nextTick(() => {
      const domSrc = wrapper.find('img').element.getAttribute('src')
      expect(domSrc).toEqual(getSrc(newSource))
      return done()
    })
  })

  test('sizes', () => {
    const sizes = wrapper.find('img').element.getAttribute('sizes')
    expect(sizes).toBe('(max-width: 500px) 500px, 900px')
  })

  test('does not register an observer for lazy loading if there is native support', () => {
    global.process.client = true
    const { loading } = global.HTMLImageElement.prototype
    global.HTMLImageElement.prototype.loading = 'exists'
    jest.resetModules()

    const NuxtImg = require('~/runtime/components/nuxt-img.vue').default
    const observer = mockObserver()
    const wrapper = mountWithImg(NuxtImg, {
      loading: 'lazy',
      width: 200,
      height: 200,
      sizes: '200,500:500,900:900',
      src
    })

    expect(observer.wasAdded).toBe(false)
    expect(observer.wasDestroyed).toBe(false)

    expect(wrapper.html()).toMatchSnapshot()

    expect(observer.wasDestroyed).toBe(false)
    HTMLImageElement.prototype.loading = loading
  })

  test('registers an observer for lazy loading if no native support for lazy loading', async () => {
    global.process.client = true
    const { loading } = global.HTMLImageElement.prototype
    delete (global.HTMLImageElement.prototype as any).loading
    jest.resetModules()

    const NuxtImg = require('~/runtime/components/nuxt-img.vue').default
    const observer = mockObserver()
    const wrapper = mountWithImg(NuxtImg, {
      loading: 'lazy',
      width: 200,
      height: 200,
      sizes: '200,500:500,900:900',
      src
    })

    expect(observer.wasAdded).toBe(true)
    expect(observer.wasDestroyed).toBe(false)

    expect(wrapper.html()).toMatchSnapshot()

    observer.triggerVisibility()

    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.destroy()
    expect(observer.wasDestroyed).toBe(true)
    HTMLImageElement.prototype.loading = loading
  })
})
