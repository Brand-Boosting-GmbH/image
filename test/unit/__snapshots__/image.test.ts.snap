// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders simple image Matches snapshot 1`] = `"<img src=\\"/_custom/image.png\\" width=\\"200\\" height=\\"200\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_custom/image.png 500w, /_custom/image.png 900w\\">"`;

exports[`Renders simple image does not register an observer for lazy loading if there is native support 1`] = `"<img src=\\"/_custom/image.png\\" width=\\"200\\" height=\\"200\\" loading=\\"lazy\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_custom/image.png 500w, /_custom/image.png 900w\\">"`;

exports[`Renders simple image registers an observer for lazy loading if no native support for lazy loading 1`] = `"<img src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7\\" width=\\"200\\" height=\\"200\\" loading=\\"lazy\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_custom/image.png 500w, /_custom/image.png 900w\\">"`;

exports[`Renders simple image registers an observer for lazy loading if no native support for lazy loading 2`] = `"<img src=\\"/_custom/image.png\\" width=\\"200\\" height=\\"200\\" loading=\\"lazy\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_custom/image.png 500w, /_custom/image.png 900w\\">"`;
